### ISSUE TITLE
`catch-error-name` fails when try/catch is inside a promise that is caught

### ISSUE THREAD

**Issue by sindresorhus** [sindresorhus/eslint-plugin-unicorn#107]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%2440%20Funded-%2300A156.svg" />](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/107)

Discovered in https://github.com/sindresorhus/got/pull/347

Failing tests: https://github.com/sindresorhus/eslint-plugin-unicorn/commit/9f9a324255cc1faaa0d5331a6b78be0cbec9db8f

The cause is the `.catch(err => err);` part. If that is removed it correctly errors that `e` should be named `err`.

// @SamVerschueren

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $40.00)

- [<img src='https://avatars3.githubusercontent.com/u/44827199?v=4' alt='issuehunt' width=24 height=24> issuehunt](https://issuehunt.io/u/issuehunt) ($40.00)

### Submitted pull Requests
- [#227 Support nesting for catch-error-name rule](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/pull/227)
- [#653 `catch-error-name`: Add nesting tests](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/pull/653)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

**Comment by SamVerschueren** [comment]
I found out why this fails. The thing is that the rule is being disabled when we have nesting catch clauses. Let's use the failing test example.

```js
foo.then(() => {
	try {} catch (e) {}
}).catch(err => err);
```

It might not look like `try-catch` and the promise `.catch` are nested, but apparently they are. So I believe that the only way to actually solve this, is by not disabling the rule when having nested catch clauses.

I believe this was allowed because it didn't allow number postfixes etc. But now we have them. So from the documentation
```js
try {
	doSomething();
} catch (anyName) { // Nesting of catch clauses disables the rule
	try {
		doSomethingElse();
	} catch (anyOtherName) {
		// ...
	}
}
```

Would actually become
```js
try {
	doSomething();
} catch (err) {
	try {
		doSomethingElse();
	} catch (err1) {
		// ...
	}
}
```

Don't think this can be auto-fixed in one go though. As the fix statement searches if `err`, `err1, `...` exists. So the inner `try-catch` would not detect that we will have an `err` identifier after fixing. Not sure if we can keep track of this though.

So that's the reason why this fails. Not sure how to go from here.

**Comment by sindresorhus** [comment]
I rather how a working rule than auto-fix, so I'm fine with that compromise. Let's go with:

> So I believe that the only way to actually solve this, is by not disabling the rule when having nested catch clauses.

**Comment by sindresorhus** [comment]
I've disabled this rule in the default config to be able to do a new release. Can be enabled again when we fix this.

**Comment by IssuehuntBot** [comment]
@issuehuntfest has funded $40.00 to this issue. [See it on IssueHunt](https://issuehunt.io/repos/55832243/issues/107)

**Comment by doniyor2109** [comment]
@sindresorhus  Is this issue still active? If it is I can work on this issue?

**Comment by sindresorhus** [comment]
Go ahead :)

**Comment by twk-b** [comment]
works in eslint-plugin-unicorn@8 but not in @9.0 or @9.1

config: unicorn/catch-error-name: ['error', {'name': 'err'}] 
code: 

`exports.get = function (req, res) {
    myfunc(function (err, values) {
        if (err) {
           console.log(err);
           return;
        } else {
                var temp;
                try {
                    temp = JSON.parse(values.value);
                }
                catch(err2) {
                    exports.services.logger.error('catching', err2);
                    temp = values;
                }
                return temp;
        }
    });
};`
result:  error  The catch parameter should be named `err_`  unicorn/catch-error-name
expected: no error

**Comment by sindresorhus** [comment]
If anyone wants to work on this, see the previous attempt in https://github.com/sindresorhus/eslint-plugin-unicorn/pull/227. You should probably continue from there.

**Comment by fisker** [comment]
> works in eslint-plugin-unicorn@8 but not in @9.0 or @9.1
> 
> config: unicorn/catch-error-name: ['error', {'name': 'err'}]
> code:
> 
> ```js
> exports.get = function (req, res) {
>     myfunc(function (err, values) {
>         if (err) {
>            console.log(err);
>            return;
>         } else {
>                 var temp;
>                 try {
>                     temp = JSON.parse(values.value);
>                 }
>                 catch(err2) {
>                     exports.services.logger.error('catching', err2);
>                     temp = values;
>                 }
>                 return temp;
>         }
>     });
> };
> ```
> 
> result: error The catch parameter should be named `err_` unicorn/catch-error-name
> expected: no error

This case should has one error, `err2` is not allowed.
